samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -443,6 +443,18 @@ public void run() {
             long startTime = clock.nanoTime();
             task.window(coordinator);
             containerMetrics.windowNs().update(clock.nanoTime() - startTime);
+
+            // We have scenario when the task window method takes longer time than window trigger time, there is a
+            // potential for process method starvation. Ideally, we need to merge the thread pool for commit, window with
+            // process to guarantee fairness. In the interim, we want to warn users about potential starvation.
+            double lowerBoundForWindowTriggerTimeInMs =
```

Comment:
this seems not very accurate. One large window might skew the avg even the process are invoked after it. And this logic shouldn't be here on the critical path of run loop. I would suggest moving this logic to window() itself, after updating the windowNs, do a count and check there. Seems more accurate that way.